litr/src/main/java/com/linkedin/android/litr/io/Camera2MediaSource.kt

```diff
@@ -0,0 +1,253 @@
+        override fun onOpened(device: CameraDevice) {
+            initDevice(device)
+        }
+
+        override fun onDisconnected(device: CameraDevice) { }
+        override fun onError(device: CameraDevice, error: Int) {
+            val msg = when(error) {
+                ERROR_CAMERA_DEVICE -> "Fatal (device)"
+                ERROR_CAMERA_DISABLED -> "Device policy"
+                ERROR_CAMERA_IN_USE -> "Camera in use"
+                ERROR_CAMERA_SERVICE -> "Fatal (service)"
+                ERROR_MAX_CAMERAS_IN_USE -> "Maximum cameras in use"
+                else -> "Unknown"
+            }
+
+            // Notify via the Callback that an error occurred.
+            val exception = RuntimeException("Camera $cameraId error: ($error) $msg")
+            Log.e(TAG, exception.message, exception)
+            callback?.onError(exception)
+        }
+    }
+
+    private val captureStateCallback = object : CameraCaptureSession.StateCallback() {
+        override fun onConfigured(session: CameraCaptureSession) {
+            captureSession = session
+            captureRequest?.let {
+                session.setRepeatingRequest(it, object : CameraCaptureSession.CaptureCallback() {
+                    override fun onCaptureCompleted(
+                            session: CameraCaptureSession,
+                            request: CaptureRequest,
+                            result: TotalCaptureResult
+                    ) {
+                        // Notify the ExternalMediaSource when a frame is captured.
+                        onFrameAvailable()
+                    }
+                }, backgroundHandler)
+            }
+        }
+
+        override fun onClosed(session: CameraCaptureSession) {
+            // When the session has been cleanly closed, we can go ahead and stop our background
+            // Thread.
+            stopBackgroundThread()
+        }
+
+        override fun onConfigureFailed(session: CameraCaptureSession) { }
+    }
+
+    init {
+        // Add ourselves as a callback to our parent ExternalMediaSource. This will allow us to
+        // determine when an input Surface is available to draw too via the Camera.
+        setCallback(this)
+    }
+
+    /**
+     * Sets the callback that will be notified on the state of our recording session.
+     */
+    fun setCallback(callback: Callback) {
+        this.callback = callback
+    }
+
+    /**
+     * Adds an optional SurfaceHolder that can be used as a Preview surface.
+     */
+    fun addPreviewSurfaceHolder(surfaceHolder: SurfaceHolder) {
+        surfaceHolder.addCallback(previewSurfaceHolderListener)
+    }
+
+    /**
+     * Opens the configured CameraDevice so that it's ready for recording. Once available, the
+     * configured Callback will be notified via onDeviceReady. Once this has done, the consumer is
+     * able to start the Transcode session.
+     */
+    @SuppressLint("MissingPermission")
```

Comment:
Can we use `@RequiresPermission` (https://developer.android.com/reference/androidx/annotation/RequiresPermission) here?

Reply:
I think we'd need to use both, right?
 - The `RequiresPermission` reports that the caller needs the permission
 - The `SuppressLint` avoids Android Studio trying to force that we add the permission to the litr manifest (instead requiring people who want/need to use the source to make sure they add it)

Reply:
Actually, scrap that, you're right (and I made a typo which is why it failed locally)